### DISCUSSION TITLE
Can web-tree-sitter actually do async parsing?

### DISCUSSION THREAD

**Discussion by savetheclocktower** [tree-sitter/tree-sitter#2118] [Q&A] [answered]
This isn't a huge deal because synchronous parsing is very quick, but it'd be good to know if it works. There's no coverage for async parsing [in the tests](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/test/parser-test.js).

I'm curious about this because `web-tree-sitter`'s `Parser` class exposes `setTimeoutMicros` and `getTimeoutMicros` methods, and setting a timeout value definitely does have an effect in my testing, but the effect is that the parsing will _fail outright_ when a timeout value is set.

Based on `node-tree-sitter` and its usage in Atom, I'd expect that the return value of `Parser#parse` would be a `Tree` whenever the parsing can be finished before the timeout, and a `Promise` whenever the parsing exceeds the timeout. Instead, in the latter case, I get an opaque `Error: Parsing failed` message.

If I set a logger via `setLogger` before the parse, I can see that the parse job progresses as usual before abruptly stopping without any obvious error, supporting my theory that the parsing stops when the timeout is exceeded.

**Comment by savetheclocktower** [comment, marked as answer]
For posterity, the answer is that it works roughly as described here:

https://github.com/tree-sitter/tree-sitter/blob/c6d663f3936aea4fb9a12e093fbaa25e762fa902/lib/include/tree_sitter/api.h#L243-L247

My mistake was assuming it'd work like the node bindings. Instead it seems to parse as much as it can in the time given, then throw an exception, and it falls to you to resume the work when/if you see fit by calling `parse` again with the same arguments.

**Reply by ahlinc** [reply]
Yeah, you can organize the parser polling in this way. I hope someday to try to implement an idea of additional push based API that I assume would provide much more control over all parsing without any downsides.
